### DISCUSSION TITLE
Representation of float values in a list

### DISCUSSION THREAD

**Discussion by AbhaMoitra** [SemanticApplicationDesignLanguage/sadl#790] [Q&A] [answered]
I am working with a list of floats and value like 2.4 is being converted to 2.4000000953674316 etc. Here is the code I am using

uri "http://sadl.org/test.sadl" alias test.

Object is a class
    described by prop1 with values of type float.

Rule MinValue
if x is a Object
   and list1 is list(x, prop1)
   and lstr is listToString(list1)
   and minValue is min(list1)
then print("list and min value are: ",minValue, lstr).
   
obj1 is a Object
	with prop1 2.4
	with prop1 3.4.    

Ask: "select * where {?x ?y ?z} limit 1".

**The output in the console is as follows**

Inference of 'C:\Sadl3Workspace21\TestList\Test.sadl' requested.
print: list and min value are:  2.4000000953674316 {2.4000000953674316, 3.4000000953674316} 
Reasoner family: Jena-Based (com.naturalsemanticsllc.sadl.reasoner.JenaAugmentedReasonerPlugin), version date 12 April 2021.
Query: select * where {?x ?y ?z} limit 1
     "x","y","z"
     "0a8ef5b3-6645-415b-aec7-175ef8dcf293(blank node)","rest","027ac450-74bd-4531-89b0-5f991b2f3d9b(blank node)"

Is there some way I can preserve 2.4 value as is?

**Comment by crapo** [comment, marked as answer]
The "list" built-in function converts floats to doubles. I don't remember why this is done, but I think there's a reason. The conversion uses the java.lang.Float class:
'''((Float)v).doubleValue()'''
For the float 2.4, this returns the value "2.4000000953674316".

**Reply by AbhaMoitra** [reply]
That helped; I used double instead of float and it kept 2.4 as 2.4.

**Comment by AbhaMoitra** [comment]
That helped; I used double instead of float and that keeps 2.4 as 2.4.

**Reply by crapo** [reply]
I tried to go back through the history and see why that was done. I was unsuccessful in that endeavor, at least so far. I think it had to do with the assumption of a float when a double is needed? We could change that code to not create double from the float and see what happens....

**Reply by crapo** [reply]
@AbhaMoitra , BTW, what did you do specifically to "use double instead of float"?

**Reply by AbhaMoitra** [reply]
I just changed float to double.

![image](https://user-images.githubusercontent.com/28540315/133608435-2c2738ab-db1f-4d38-80c2-ecd7f0db8671.png)

**Comment by crapo** [comment]
@AbhaMoitra , I think that the relevant history of List.java is in openge. It was part of the proprietary Jena reasoner for GE before being moved to SADL 3 on github. I don't have access to that repo. It would have been in the project for that reasoner I think. If you want to look and see if you can find the history that might be helpful in terms of understanding why it is implemented as it is. Or we could have a Teams meeting and I could help you find the relevant history. If it's worth it to you.

**Reply by AbhaMoitra** [reply]
@crapo : If you think it may be helpful to look at the history then we could do a Teams meeting, I am personally not interested in the history.

**Comment by crapo** [comment]
@AbhaMoitra , this conversion of float to double is a pattern across many built-ins. There is a reason, but it escapes me. Unless you wish to pursue it I think we will not investigate further at this time and leave it as is.

**Reply by AbhaMoitra** [reply]
Yes agreed, let us leave it as is.
